### PR TITLE
fix mypy typing errors in pytorch_lightning/strategies/horovod.py

### PR DESCRIPTION
**Pull request by Cyprien-Ricque**
## What does this PR do?

Fixes mypy typing errors in pytorch_lightning/strategies/horovod.py in https://github.com/Lightning-AI/lightning/issues/13445.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/strategies/horovod.py [hunk 1]
```diff
@@ -129,27 +133,33 @@ def _unpack_lightning_optimizer(opt):
...
+    def join(self) -> None:
         if self.root_device.type == "cuda":
             hvd.join(self.local_rank)
         else:
             hvd.join()
 
-    def reduce(self, tensor, group: Optional[Any] = None, reduce_op: Optional[Union[ReduceOp, str]] = "mean"):
+    def reduce(
+        self,
+        tensor: Union[Any, Tensor],
```
[the remark this answers is not in the document]
Comment (Borda) [reply src/pytorch_lightning/strategies/horovod.py:159]: yes `Union[Any, Tensor]` -> `Any` but it is confusing as `tensor` argument can be any

[the remark this answers is not in the document]
Comment (akihironitta) [reply src/pytorch_lightning/strategies/horovod.py:159]: I left a similar comment in a PR quite a while ago, but I remember we concluded that we'd continue to have types like this instead of merging them to `Any`. (cc @justusschock if I remember it correctly)

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/strategies/hivemind.py [hunk 2]
```diff
@@ -315,6 +315,7 @@ def __init__(self, optimizer: "hivemind.Optimizer", scheduler: _LRScheduler) ->
         # implemented custom logic which we would not want to call on destruction of the `HiveMindScheduler`
         self.__dict__ = {k: v for k, v in scheduler.__dict__.items() if k not in ("step", "__del__")}
 
+        self.base_lrs = scheduler.base_lrs
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/hivemind.py:318]: This should be already set from the `__dict__` override 2 lines above

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/strategies/hivemind.py (hunk 2 above)
Comment (otaj) [reply src/pytorch_lightning/strategies/hivemind.py:318]: That is true, but I think mypy does not pick it up since it's set by `__dict__` override

Comment (justusschock) [reply src/pytorch_lightning/strategies/hivemind.py:318]: But then we probably should just add stubs before the init instead of doing that again

File: src/pytorch_lightning/strategies/horovod.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/strategies/horovod.py:159]: The advantage that way is that actually most IDEs nowadays parse the typehints and that way you do get autocompletion for tensors which in many cases is the correct type instead of just doing Any, where the IDE doesn't know what to suggest. So yeah, technically this is redundant, but it really helps with delevoper experience

File: src/pytorch_lightning/strategies/horovod.py [hunk 3]
```diff
@@ -111,8 +113,10 @@ def _unpack_lightning_optimizer(opt):
         lr_scheduler_configs = self.lr_scheduler_configs
         for config in lr_scheduler_configs:
             scheduler = config.scheduler
+            assert isinstance(scheduler, _LRScheduler), "ReduceLROnPlateau does not have attribute 'base_lr'"
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/horovod.py:116]: The assertion does not relate to the attribute base_lr. 
The class of the scheduler is also not ReduceLROnPlateau in general. 
I suggest you remove the comment

File: src/pytorch_lightning/strategies/horovod.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/strategies/horovod.py:159]: I am in favor of having Any, Tensor

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/strategies/horovod.py (hunk 3 above)
Comment (Cyprien-Ricque) [reply src/pytorch_lightning/strategies/horovod.py:116]: Sorry, indeed I made a typo, it is ```base_lrs```.
According to the type of ```scheduler``` being ```Union[_LRScheduler, ReduceLROnPlateau]``` If the type is not ```_LRScheduler``` then it is ```ReduceLROnPlateau```, right? 
Do you mean that it is also used with other schedulers not listed is that ```Union```?
In that case shouldn't we update the ```scheduler``` type in the data class ```LRSchedulerConfig```?

File: src/pytorch_lightning/strategies/hivemind.py (hunk 2 above)
Comment (Cyprien-Ricque) [reply src/pytorch_lightning/strategies/hivemind.py]: I did it. Thank you for the advice!
